File: android/src/main/kotlin/com/melodi/core/MelodiCorePlugin.kt
package com.melodi.core

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject

class MelodiCorePlugin: FlutterPlugin, MethodCallHandler {
    private var channel: MethodChannel? = null
    private var context: Context? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "melodi/core")
        channel?.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onDetachedFromEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        context = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initialize" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val storageRoot = context?.getExternalFilesDir(null)?.absolutePath ?: context?.filesDir?.absolutePath ?: ""
                val initResult = MelodiCoreJNI.init(storageRoot)
                if (initResult == 0) {
                    result.success(mapOf(
                        "version" to MelodiCoreJNI.version(),
                        "apiVersion" to MelodiCoreJNI.apiVersion()
                    ))
                } else {
                    result.error("INIT_FAILED", "Failed to initialize MelodiCore", null)
                }
            }
            "ping" -> {
                result.success(MelodiCoreJNI.ping())
            }
            "search" -> {
                val json = JSONObject(call.arguments as? Map<String, Any> ?: emptyMap()).toString()
                val response = MelodiCoreJNI.search(json)
                result.success(response)
            }
            "match" -> {
                val json = JSONObject(call.arguments as? Map<String, Any> ?: emptyMap()).toString()
                val response = MelodiCoreJNI.match(json)
                result.success(response)
            }
            "resolve" -> {
                val json = JSONObject(call.arguments as? Map<String, Any> ?: emptyMap()).toString()
                val response = MelodiCoreJNI.resolve(json)
                result.success(response)
            }
            "download" -> {
                val json = JSONObject(call.arguments as? Map<String, Any> ?: emptyMap()).toString()
                val response = MelodiCoreJNI.download(json)
                result.success(response)
            }
            "cancelDownload" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val jobId = args["jobId"] as? String ?: ""
                MelodiCoreJNI.cancelDownload(jobId)
                result.success(null)
            }
            "getDownloadStatus" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val jobId = args["jobId"] as? String ?: ""
                val response = MelodiCoreJNI.getDownloadStatus(jobId)
                result.success(response)
            }
            "listDownloads" -> {
                val response = MelodiCoreJNI.listDownloads()
                result.success(response)
            }
            "installExtension" -> {
                val json = JSONObject(call.arguments as? Map<String, Any> ?: emptyMap()).toString()
                val response = MelodiCoreJNI.installExtension(json)
                result.success(response)
            }
            "uninstallExtension" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val id = args["id"] as? String ?: ""
                MelodiCoreJNI.uninstallExtension(id)
                result.success(null)
            }
            "enableExtension" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val id = args["id"] as? String ?: ""
                val enabled = args["enabled"] as? Boolean ?: false
                MelodiCoreJNI.enableExtension(id, enabled)
                result.success(null)
            }
            "listExtensions" -> {
                val response = MelodiCoreJNI.listExtensions()
                result.success(response)
            }
            "getExtension" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val id = args["id"] as? String ?: ""
                val response = MelodiCoreJNI.getExtension(id)
                result.success(response)
            }
            "checkExtensionHealth" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val id = args["id"] as? String ?: ""
                val response = MelodiCoreJNI.checkExtensionHealth(id)
                result.success(response)
            }
            "updateExtension" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val id = args["id"] as? String ?: ""
                val response = MelodiCoreJNI.updateExtension(id)
                result.success(response)
            }
            "updateAllExtensions" -> {
                val response = MelodiCoreJNI.updateAllExtensions()
                result.success(response)
            }
            "addRepository" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val url = args["url"] as? String ?: ""
                MelodiCoreJNI.addRepository(url)
                result.success(null)
            }
            "removeRepository" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val url = args["url"] as? String ?: ""
                MelodiCoreJNI.removeRepository(url)
                result.success(null)
            }
            "listRepositories" -> {
                val response = MelodiCoreJNI.listRepositories()
                result.success(response)
            }
            "readMetadata" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val filePath = args["filePath"] as? String ?: ""
                val response = MelodiCoreJNI.readMetadata(filePath)
                result.success(response)
            }
            "writeMetadata" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val filePath = args["filePath"] as? String ?: ""
                val tags = args["tags"] as? Map<String, Any> ?: emptyMap()
                val tagsJson = JSONObject(tags).toString()
                MelodiCoreJNI.writeMetadata(filePath, tagsJson)
                result.success(null)
            }
            "embedCoverArt" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val filePath = args["filePath"] as? String ?: ""
                val imageData = args["imageData"] as? List<Int> ?: emptyList()
                val mimeType = args["mimeType"] as? String ?: ""
                val byteArray = imageData.map { it.toByte() }.toByteArray()
                MelodiCoreJNI.embedCoverArt(filePath, byteArray, mimeType)
                result.success(null)
            }
            "embedLyrics" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val filePath = args["filePath"] as? String ?: ""
                val lyrics = args["lyrics"] as? String ?: ""
                MelodiCoreJNI.embedLyrics(filePath, lyrics)
                result.success(null)
            }
            "extractLyrics" -> {
                val args = call.arguments as? Map<String, Any> ?: emptyMap()
                val filePath = args["filePath"] as? String ?: ""
                val response = MelodiCoreJNI.extractLyrics(filePath)
                result.success(response)
            }
            "getStats" -> {
                val response = MelodiCoreJNI.getStats()
                result.success(response)
            }
            else -> {
                result.notImplemented()
            }
        }
    }
}

object MelodiCoreJNI {
    @JvmStatic
    external fun init(storageRoot: String): Int

    @JvmStatic
    external fun version(): String

    @JvmStatic
    external fun apiVersion(): String

    @JvmStatic
    external fun ping(): String

    @JvmStatic
    external fun search(requestJson: String): String

    @JvmStatic
    external fun match(requestJson: String): String

    @JvmStatic
    external fun resolve(requestJson: String): String

    @JvmStatic
    external fun download(requestJson: String): String

    @JvmStatic
    external fun cancelDownload(jobId: String)

    @JvmStatic
    external fun getDownloadStatus(jobId: String): String

    @JvmStatic
    external fun listDownloads(): String

    @JvmStatic
    external fun installExtension(requestJson: String): String

    @JvmStatic
    external fun uninstallExtension(id: String)

    @JvmStatic
    external fun enableExtension(id: String, enabled: Boolean)

    @JvmStatic
    external fun listExtensions(): String

    @JvmStatic
    external fun getExtension(id: String): String

    @JvmStatic
    external fun checkExtensionHealth(id: String): String

    @JvmStatic
    external fun updateExtension(id: String): String

    @JvmStatic
    external fun updateAllExtensions(): String

    @JvmStatic
    external fun addRepository(url: String)

    @JvmStatic
    external fun removeRepository(url: String)

    @JvmStatic
    external fun listRepositories(): String

    @JvmStatic
    external fun readMetadata(filePath: String): String

    @JvmStatic
    external fun writeMetadata(filePath: String, tagsJson: String)

    @JvmStatic
    external fun embedCoverArt(filePath: String, imageData: ByteArray, mimeType: String)

    @JvmStatic
    external fun embedLyrics(filePath: String, lyrics: String)

    @JvmStatic
    external fun extractLyrics(filePath: String): String

    @JvmStatic
    external fun getStats(): String

    companion object {
        init {
            System.loadLibrary("melodi_core")
        }
    }
}

File: ios/Classes/MelodiCorePlugin.m
#import <Flutter/Flutter.h>
#import <Foundation/Foundation.h>

extern int MelodiCore_Init(const char* storage_root);
extern const char* MelodiCore_Version();
extern const char* MelodiCore_APIVersion();
extern const char* MelodiCore_Ping();
extern void MelodiCore_Shutdown();
extern const char* MelodiCore_Search(const char* request_json);
extern const char* MelodiCore_Match(const char* request_json);
extern const char* MelodiCore_Resolve(const char* request_json);
extern const char* MelodiCore_Download(const char* request_json);
extern void MelodiCore_CancelDownload(const char* job_id);
extern const char* MelodiCore_GetDownloadStatus(const char* job_id);
extern const char* MelodiCore_ListDownloads();
extern const char* MelodiCore_InstallExtension(const char* request_json);
extern void MelodiCore_UninstallExtension(const char* id);
extern void MelodiCore_EnableExtension(const char* id, int enabled);
extern const char* MelodiCore_ListExtensions();
extern const char* MelodiCore_GetExtension(const char* id);
extern const char* MelodiCore_CheckExtensionHealth(const char* id);
extern const char* MelodiCore_UpdateExtension(const char* id);
extern const char* MelodiCore_UpdateAllExtensions();
extern void MelodiCore_AddRepository(const char* url);
extern void MelodiCore_RemoveRepository(const char* url);
extern const char* MelodiCore_ListRepositories();
extern const char* MelodiCore_ReadMetadata(const char* file_path);
extern void MelodiCore_WriteMetadata(const char* file_path, const char* tags_json);
extern void MelodiCore_EmbedCoverArt(const char* file_path, const char* image_data, int image_len, const char* mime_type);
extern void MelodiCore_EmbedLyrics(const char* file_path, const char* lyrics);
extern const char* MelodiCore_ExtractLyrics(const char* file_path);
extern const char* MelodiCore_GetStats();

@interface MelodiCorePlugin : NSObject <FlutterPlugin>
@end

@implementation MelodiCorePlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"melodi/core"
            binaryMessenger:[registrar messenger]];
  MelodiCorePlugin* instance = [[MelodiCorePlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"initialize" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* storageRoot = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES).firstObject stringByAppendingPathComponent:@"melodi"];
    const char* cStorageRoot = [storageRoot UTF8String];
    int initResult = MelodiCore_Init(cStorageRoot);
    if (initResult == 0) {
      result(@{
        @"version": [NSString stringWithUTF8String:MelodiCore_Version()],
        @"apiVersion": [NSString stringWithUTF8String:MelodiCore_APIVersion()],
      });
    } else {
      result(FlutterError(code: "INIT_FAILED", message: "Failed to initialize MelodiCore", details: nil));
    }
  } else if ([@"ping" isEqualToString:call.method]) {
    result([NSString stringWithUTF8String:MelodiCore_Ping()]);
  } else if ([@"search" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:args options:0 error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    const char* response = MelodiCore_Search([jsonString UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"match" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:args options:0 error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    const char* response = MelodiCore_Match([jsonString UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"resolve" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:args options:0 error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    const char* response = MelodiCore_Resolve([jsonString UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"download" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:args options:0 error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    const char* response = MelodiCore_Download([jsonString UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"cancelDownload" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* jobId = args[@"jobId"];
    MelodiCore_CancelDownload([jobId UTF8String]);
    result(nil);
  } else if ([@"getDownloadStatus" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* jobId = args[@"jobId"];
    const char* response = MelodiCore_GetDownloadStatus([jobId UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"listDownloads" isEqualToString:call.method]) {
    const char* response = MelodiCore_ListDownloads();
    result([NSString stringWithUTF8String:response]);
  } else if ([@"installExtension" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:args options:0 error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    const char* response = MelodiCore_InstallExtension([jsonString UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"uninstallExtension" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* id = args[@"id"];
    MelodiCore_UninstallExtension([id UTF8String]);
    result(nil);
  } else if ([@"enableExtension" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* id = args[@"id"];
    NSNumber* enabled = args[@"enabled"];
    MelodiCore_EnableExtension([id UTF8String], [enabled intValue]);
    result(nil);
  } else if ([@"listExtensions" isEqualToString:call.method]) {
    const char* response = MelodiCore_ListExtensions();
    result([NSString stringWithUTF8String:response]);
  } else if ([@"getExtension" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* id = args[@"id"];
    const char* response = MelodiCore_GetExtension([id UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"checkExtensionHealth" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* id = args[@"id"];
    const char* response = MelodiCore_CheckExtensionHealth([id UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"updateExtension" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* id = args[@"id"];
    const char* response = MelodiCore_UpdateExtension([id UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"updateAllExtensions" isEqualToString:call.method]) {
    const char* response = MelodiCore_UpdateAllExtensions();
    result([NSString stringWithUTF8String:response]);
  } else if ([@"addRepository" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* url = args[@"url"];
    MelodiCore_AddRepository([url UTF8String]);
    result(nil);
  } else if ([@"removeRepository" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* url = args[@"url"];
    MelodiCore_RemoveRepository([url UTF8String]);
    result(nil);
  } else if ([@"listRepositories" isEqualToString:call.method]) {
    const char* response = MelodiCore_ListRepositories();
    result([NSString stringWithUTF8String:response]);
  } else if ([@"readMetadata" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* filePath = args[@"filePath"];
    const char* response = MelodiCore_ReadMetadata([filePath UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"writeMetadata" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* filePath = args[@"filePath"];
    NSDictionary* tags = args[@"tags"];
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:tags options:0 error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    MelodiCore_WriteMetadata([filePath UTF8String], [jsonString UTF8String]);
    result(nil);
  } else if ([@"embedCoverArt" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* filePath = args[@"filePath"];
    NSData* imageData = args[@"imageData"];
    NSString* mimeType = args[@"mimeType"];
    MelodiCore_EmbedCoverArt([filePath UTF8String], [imageData bytes], (int)[imageData length], [mimeType UTF8String]);
    result(nil);
  } else if ([@"embedLyrics" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* filePath = args[@"filePath"];
    NSString* lyrics = args[@"lyrics"];
    MelodiCore_EmbedLyrics([filePath UTF8String], [lyrics UTF8String]);
    result(nil);
  } else if ([@"extractLyrics" isEqualToString:call.method]) {
    NSDictionary* args = call.arguments;
    NSString* filePath = args[@"filePath"];
    const char* response = MelodiCore_ExtractLyrics([filePath UTF8String]);
    result([NSString stringWithUTF8String:response]);
  } else if ([@"getStats" isEqualToString:call.method]) {
    const char* response = MelodiCore_GetStats();
    result([NSString stringWithUTF8String:response]);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end